### DISCUSSION TITLE
Tree shaking in next js for library with module as common js

### DISCUSSION THREAD

**Discussion by ShubhamTheReactGeek** [vercel/next.js#43063] [Help]
I'm importing one of our library which is transpiling it's code to commonjs, next js is not performing tree shaking on this library. 
Please help.

**Comment by icyJoseph** [comment]
Hi,

Tree shaking, or rather side-effect analysis and dead code-elimination are tricky.

Often it is Webpack who can't really conclude if some code is to be dropped or included. Most likely because it is unclear whether the presence of this code has any impact on functionality (side-effect).

Infamously: 

```tsx
import {Icon} from 'some-icon-lib'
```

Can end up including all icons in the library! So we are forced to do:

```tsx
import {Icon} from 'some-icon-lib/Icon'
```

Or some other flavor/combination of that. At the end of the day, we are in charge, so we ought to use bundle analyzers and spend time pruning unwanted code.

If it is not working out of the box, then you might have to guide Webpack's hand by making the import paths more explicit and such. 

Do you see any possibility to try and narrow the import paths?

**Reply by ShubhamTheReactGeek** [reply]
Hi,
Thanks for replying. 

I have mentioned `sideEffects:false` in package.json of my library. So webpack should know that there is no sideEffects in this library so it can remove unused code from bundle and perform the treeshaking. 
As you mentioned If I am only importing Icon from library, why there is need to bundle the whole library ?
Do you have any idea how to make this happen in Next JS ?

Thanks
